File: server/src/main/java/au/com/codeka/warworlds/server/concurrency/TaskRunner.java
package au.com.codeka.warworlds.server.concurrency;

import java.util.Timer;
import java.util.TimerTask;

/**
 * This is a class for running tasks on various threads. You can run a task on any thread defined
 * in {@link Threads}.
 */
public class TaskRunner {
  public static TaskRunner i = new TaskRunner();

  private Timer timer;

  private TaskRunner() {
    ThreadPool backgroundThreadPool = new ThreadPool(
        Threads.BACKGROUND,
        2500 /* maxQueuedItems */,
        10 /* minThreads */,
        50 /* maxThreads */,
        5000 /* keepAliveMs */);
    Threads.BACKGROUND.setThreadPool(backgroundThreadPool);

    timer = new Timer("Timer");
  }

  public void runTask(Runnable runnable, Threads thread) {
    thread.runTask(runnable);
  }

  /** Run a task after the given delay. */
  public void runTask(final Runnable runnable, final Threads thread, int delayMs) {
    if (delayMs == 0) {
      runTask(runnable, thread);
    } else {
      timer.schedule(new TimerTask() {
        @Override
        public void run() {
          runTask(runnable, thread);
        }
      }, delayMs);
    }
  }
}

File: server/src/main/java/au/com/codeka/warworlds/server/concurrency/TaskRunner.kt
package au.com.codeka.warworlds.server.concurrency

import java.util.*

/**
 * This is a class for running tasks on various threads. You can run a task on any thread defined
 * in [Threads].
 */
class TaskRunner private constructor() {
  private val timer: Timer
  fun runTask(runnable: Runnable, thread: Threads) {
    thread.runTask(runnable)
  }

  /** Run a task after the given delay.  */
  fun runTask(runnable: Runnable, thread: Threads, delayMs: Int) {
    if (delayMs == 0) {
      runTask(runnable, thread)
    } else {
      timer.schedule(object : TimerTask() {
        override fun run() {
          runTask(runnable, thread)
        }
      }, delayMs.toLong())
    }
  }

  companion object {
    var i = TaskRunner()
  }

  init {
    val backgroundThreadPool = ThreadPool(
        Threads.BACKGROUND,
        2500 /* maxQueuedItems */,
        10 /* minThreads */,
        50 /* maxThreads */,
        5000 /* keepAliveMs */)
    Threads.BACKGROUND.setThreadPool(backgroundThreadPool)
    timer = Timer("Timer")
  }
}